src/libproc_macro/bridge.rs

```diff
@@ -0,0 +1,856 @@
+                $(type $name = super::$name;)*
+            }
+
+            $(impl<S: $name> $name for Handlify<S> {
+                $(fn $method(&self, $($arg: $arg_ty),*) $(-> $ret_ty)? {
+                    let storage = &mut *self.storage.borrow_mut();
+                    let r = {
+                        handlify_args!(storage; $($arg),*);
+                        $name::$method(&self.concrete, $($arg),*)
+                    };
+                    <_>::from_data(r, storage)
+                })*
+            })*
+        }
+    }
+    with_api!(Self, self_, define_handlify_impls);
+
+    impl Expand1 {
+        pub fn run<S: Server>(
+            &self,
+            server: S,
+            input: S::TokenStream,
+        ) -> thread::Result<S::TokenStream> {
+            let server = Handlify::new(self.counters, server);
+            let input = super::TokenStream::from_data(input, &mut server.storage.borrow_mut());
+            let output = (self.run)(&(&server as &super::Server), input);
+            output.map(|output| output.to_data(&mut server.storage.borrow_mut()))
+        }
+    }
+
+    impl Expand2 {
+        pub fn run<S: Server>(
+            &self,
+            server: S,
+            input: S::TokenStream,
+            input2: S::TokenStream
+        ) -> thread::Result<S::TokenStream> {
+            let server = Handlify::new(self.counters, server);
+            let input = super::TokenStream::from_data(input, &mut server.storage.borrow_mut());
+            let input2 = super::TokenStream::from_data(input2, &mut server.storage.borrow_mut());
+            let output = (self.run)(&(&server as &super::Server), input, input2);
+            output.map(|output| output.to_data(&mut server.storage.borrow_mut()))
+        }
+    }
+
+    pub trait Registry {
+        fn register_custom_derive(
+            &mut self,
+            trait_name: &str,
+            expand: Expand1,
+            attributes: &[&'static str],
+        );
+
+        fn register_attr_proc_macro(
+            &mut self,
+            name: &str,
+            expand: Expand2,
+        );
+
+        fn register_bang_proc_macro(
+            &mut self,
+            name: &str,
+            expand: Expand1,
+        );
+    }
+}
+
+trait HandleFromData<S: server::Types> {
+    type Data;
+    fn from_data(x: Self::Data, storage: &mut Storage<S>) -> Self;
+}
+
+trait HandleToData<'s, S: server::Types> {
+    type Data;
+    fn to_data(self, storage: &'s mut Storage<S>) -> Self::Data;
+}
+
+// FIXME(eddyb) achieve ABI compatibility for these types.
+impl<S: server::Types, H: HandleFromData<S>> HandleFromData<S> for Option<H> {
+    type Data = Option<H::Data>;
+    fn from_data(x: Self::Data, storage: &mut Storage<S>) -> Self {
+        x.map(|x| H::from_data(x, storage))
+    }
+}
+
+macro_rules! passthrough_from_and_to_data {
+    ($($ty:ty),*) => {
+        $(
+            impl<S: server::Types> HandleFromData<S> for $ty {
+                type Data = Self;
```

Comment:
So just to make sure I understand, `Data` here is what's actually crossing the ABI boundary and `HandleFromData` is basically "how do I create myself from an ABI value", right?

Reply:
Yes, although it's very important and the data never leaves the server, and only the handle crosses the bridge, via the C ABI (or in the future, serialization).

With serialization, we'd notably have two transformations, one would still be the handlification that's done here, while the second one would be writing/reading into/from some sort of buffer (which for handlified data, would be just the integer inside the handle).